File: InfixPrefixPostfixExp/InfixToPostfix/inToPost.py
class Solution:
    def solve(self, val1, op, val2):
        return val1 + val2 + op

    def checkOperand(self, term):
        if 'a' <= term <= 'z':
            return True
        elif 'A' <= term <= 'Z':
            return True
        elif '0' <= term <= '9':
            return True
        return False

    def prio(self, op):
        if op == '+' or op == '-':
            return 1
        elif op == '*' or op == '/':
            return 2
        return 3

    def infixToPostfix(self, s):
        operand = []
        operators = []

        for term in s:

            # Operand
            if self.checkOperand(term):
                operand.append(term)

            # Opening bracket
            elif term == '(':
                operators.append(term)

            # No operator present
            elif not operators:
                operators.append(term)

            # Closing bracket
            elif term == ')':
                while operators and operators[-1] != '(':
                    val2 = operand.pop()
                    val1 = operand.pop()

                    op = operators.pop()

                    res = self.solve(val1, op, val2)
                    operand.append(res)

                # Remove '('
                operators.pop()

            # '(' is on top
            elif operators[-1] == '(':
                operators.append(term)

            # Current operator has higher priority
            elif self.prio(operators[-1]) < self.prio(term):
                operators.append(term)

            # Equal/lower priority
            else:
                while (operators and
                       operators[-1] != '(' and
                       (self.prio(term) < self.prio(operators[-1]) or
                        (self.prio(term) == self.prio(operators[-1]) and term != '^'))):

                    val2 = operand.pop()
                    val1 = operand.pop()

                    op = operators.pop()

                    res = self.solve(val1, op, val2)
                    operand.append(res)

                operators.append(term)

        # Remaining operators
        while operators:
            val2 = operand.pop()
            val1 = operand.pop()

            op = operators.pop()

            res = self.solve(val1, op, val2)
            operand.append(res)

        return operand[-1]

File: InfixPrefixPostfixExp/InfixToPostfix/inToPost.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    string solve(string val1, char op, string val2){
        string ans = "";
        ans += val1;
        ans += val2;
        ans += op;
        return ans;
    }
    bool checkOperand(char term){
        if(term>='a' && term<='z')return true;
        else if(term>='A' && term<='Z')return true;
        else if(term>='0' && term<='9')return true;
        return false;
    }
    int prio(char op){
        if(op == '+' || op == '-')return 1;
        else if(op == '*' || op == '/')return 2;
        return 3;
    }
    string infixToPostfix(string& s) {
        // code here
        int n = s.size();
        stack<string>operand;
        stack<char>operators;
        for(int i=0;i<n;i++){
            char term = s[i];
            if(checkOperand(term))operand.push(string(1, term));
            else if(term=='(')operators.push(term);
            else if(!operators.size()) operators.push(term);
            else if(term == ')'){
                while(!operators.empty() && operators.top() != '('){
                    string val2 = operand.top();
                    operand.pop();
                    string val1 = operand.top();
                    operand.pop();
                    char op = operators.top();
                    operators.pop();
                    string res = solve(val1,op,val2);
                    operand.push(res);
                }
                operators.pop();
            }
            else if(operators.size() && operators.top() == '(')operators.push(term);
            else if(operators.size() && prio(operators.top()) < prio(term))operators.push(term);
            else{
                while(operators.size() && operators.top() != '(' && (prio(term) < prio(operators.top()) || (prio(term) == prio(operators.top()) && term != '^'))){
                    string val2 = operand.top();
                    operand.pop();
                    string val1 = operand.top();
                    operand.pop();
                    char op = operators.top();
                    operators.pop();
                    string res = solve(val1,op,val2);
                    operand.push(res);
                }
                operators.push(term);
            }
        }
        while(!operators.empty()){
            string val2= operand.top();
            operand.pop();
            string val1 = operand.top();
            operand.pop();
            char op = operators.top();
            operators.pop();
            string res = solve(val1,op,val2);
            operand.push(res);
        }
        return operand.top();
    }
};
